<a href="https://colab.research.google.com/github/genge-trw/Interactive-Data-Storyteller/blob/main/nb/GPT_OSS_MXFP4_(20B)-Inference.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [18]:
%%capture
import os, importlib.util
!pip install --upgrade -qqq uv

# Install unsloth and unsloth_zoo. Unsloth manages its own dependencies for transformers, trl, peft, accelerate.
!uv pip install -qqq \
    "unsloth[base] @ git+https://github.com/unslothai/unsloth" \
    "unsloth_zoo[base] @ git+https://github.com/unslothai/unsloth-zoo"

# Explicitly install/upgrade triton to ensure compatibility with transformers and unsloth.
!pip install --upgrade -qqq triton

In [20]:
from unsloth import FastLanguageModel
import torch

# 4bit pre quantized models we support for 4x faster downloading + no OOMs.
fourbit_models = [
    "unsloth/gpt-oss-20b-unsloth-bnb-4bit", # 20B model using bitsandbytes 4bit quantization
    "unsloth/gpt-oss-120b-unsloth-bnb-4bit",
    "unsloth/gpt-oss-20b", # 20B model using MXFP4 format
    "unsloth/gpt-oss-120b",
] # More models at https://huggingface.co/unsloth

model, tokenizer = FastLanguageModel.from_pretrained(
    model_name = "unsloth/gpt-oss-20b-unsloth-bnb-4bit",
    dtype = None, # None for auto detection
    max_seq_length = 4096, # Choose any for long context!
    load_in_4bit = True,  # 4 bit quantization to reduce memory
    full_finetuning = True, # [NEW!] We have full finetuning now!
    # token = "hf_...", # use one if using gated models
)

Unsloth: You selected full finetuning support, but 4bit / 8bit is enabled - disabling LoRA / QLoRA.
==((====))==  Unsloth 2025.11.3: Fast Gpt_Oss patching. Transformers: 4.56.2.
   \\   /|    Tesla T4. Num GPUs = 1. Max memory: 14.741 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.9.0+cu126. CUDA: 7.5. CUDA Toolkit: 12.6. Triton: 3.5.0
\        /    Bfloat16 = FALSE. FA [Xformers = None. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!
Unsloth: Using float16 precision for gpt_oss won't work! Using float32.
Unsloth: Using bfloat16 full finetuning which cuts memory usage by 50%.
To enable float32 training, use `float32_mixed_precision = True` during FastLanguageModel.from_pretrained


Loading checkpoint shards:   0%|          | 0/3 [00:00<?, ?it/s]

AttributeError: module 'triton_kernels' has no attribute 'matmul_ogs'

In [5]:
from transformers import TextStreamer

messages = [
    {"role": "user", "content": "Solve x^5 + 3x^4 - 10 = 3."},
]
inputs = tokenizer.apply_chat_template(
    messages,
    add_generation_prompt = True,
    return_tensors = "pt",
    return_dict = True,
    reasoning_effort = "low", # **NEW!** Set reasoning effort to low, medium or high
).to("cuda")

_ = model.generate(**inputs, max_new_tokens = 512, streamer = TextStreamer(tokenizer))

NameError: name 'tokenizer' is not defined

In [12]:
from transformers import TextStreamer

messages = [
    {"role": "user", "content": "Solve x^5 + 3x^4 - 10 = 3."},
]
inputs = tokenizer.apply_chat_template(
    messages,
    add_generation_prompt = True,
    return_tensors = "pt",
    return_dict = True,
    reasoning_effort = "high", # **NEW!** Set reasoning effort to low, medium or high
).to("cuda")

_ = model.generate(**inputs, max_new_tokens = 2048, streamer = TextStreamer(tokenizer))

NameError: name 'tokenizer' is not defined

In [43]:
oss = '/root/.cache/huggingface/hub/models--unsloth--gpt-oss-20b/snapshots/e220476dc09936adfed96d0451acfa3601c23bd7/'

In [24]:
#!pip install llama-cpp-python

from llama_cpp import Llama

llm = Llama.from_pretrained(
    repo_id="mradermacher/Qwen3-4B-Claude-Sonnet-4-Reasoning-Distill-Safetensor-GGUF",
    filename="Qwen3-4B-Claude-Sonnet-4-Reasoning-Distill-Safetensor.Q4_K_M.gguf",
)

./Qwen3-4B-Claude-Sonnet-4-Reasoning-Dis(…):   0%|          | 0.00/2.50G [00:00<?, ?B/s]

llama_model_loader: loaded meta data with 35 key-value pairs and 398 tensors from /root/.cache/huggingface/hub/models--mradermacher--Qwen3-4B-Claude-Sonnet-4-Reasoning-Distill-Safetensor-GGUF/snapshots/6862150716160dca2edbd80245e2c5a4ca1a2eac/./Qwen3-4B-Claude-Sonnet-4-Reasoning-Distill-Safetensor.Q4_K_M.gguf (version GGUF V3 (latest))
llama_model_loader: Dumping metadata keys/values. Note: KV overrides do not apply in this output.
llama_model_loader: - kv   0:                       general.architecture str              = qwen3
llama_model_loader: - kv   1:                               general.type str              = model
llama_model_loader: - kv   2:                               general.name str              = Qwen3 4B Claude Sonnet 4 Reasoning Di...
llama_model_loader: - kv   3:                            general.version str              = 4
llama_model_loader: - kv   4:                           general.finetune str              = Claude-Sonnet-Reasoning-Distill-Safet...
llama_mo

In [20]:

# !pip install llama-cpp-python

from llama_cpp import Llama

llm = Llama.from_pretrained(
    repo_id="nomic-ai/nomic-embed-text-v2-moe-GGUF",
    filename="nomic-embed-text-v2-moe.f32.gguf",
    n_ctx=512
)

llama_model_loader: loaded meta data with 45 key-value pairs and 142 tensors from /root/.cache/huggingface/hub/models--nomic-ai--nomic-embed-text-v2-moe-GGUF/snapshots/ffbcf4c99e5d617dda10ec8c0e9f75754b0cbb80/./nomic-embed-text-v2-moe.f32.gguf (version GGUF V3 (latest))
llama_model_loader: Dumping metadata keys/values. Note: KV overrides do not apply in this output.
llama_model_loader: - kv   0:                       general.architecture str              = nomic-bert-moe
llama_model_loader: - kv   1:                               general.type str              = model
llama_model_loader: - kv   2:                               general.name str              = nomic-embed-text-v2-moe
llama_model_loader: - kv   3:                            general.version str              = 2048
llama_model_loader: - kv   4:                       general.organization str              = Nomic Ai
llama_model_loader: - kv   5:                           general.basename str              = nomic-xlm
llama_mode

In [15]:
!pip install llama-cpp-python

from llama_cpp import Llama

llm = Llama.from_pretrained(
 repo_id="mradermacher/falcon-7b-instruct-GGUF",
 filename="falcon-7b-instruct.Q4_K_M.gguf",
 n_ctx=2048, # Explicitly set context length to match the model's n_ctx_train
)

llama_model_loader: loaded meta data with 37 key-value pairs and 196 tensors from /root/.cache/huggingface/hub/models--mradermacher--falcon-7b-instruct-GGUF/snapshots/3936b4e4b10e02658a85212d4ae7bdb425901e32/./falcon-7b-instruct.Q4_K_M.gguf (version GGUF V3 (latest))
llama_model_loader: Dumping metadata keys/values. Note: KV overrides do not apply in this output.
llama_model_loader: - kv   0:                       general.architecture str              = falcon
llama_model_loader: - kv   1:                               general.type str              = model
llama_model_loader: - kv   2:                               general.name str              = Falcon 7b Instruct
llama_model_loader: - kv   3:                           general.finetune str              = instruct
llama_model_loader: - kv   4:                           general.basename str              = falcon
llama_model_loader: - kv   5:                         general.size_label str              = 7B
llama_model_loader: - kv   6:  

In [26]:

#!apt install cmake -y
!git clone https://github.com/ggml-org/llama.cpp.git
%cd llama.cpp/
!cmake -B build
!cmake --build build -j 2 --target llama-quantize llama-cli llama-run

Cloning into 'llama.cpp'...
remote: Enumerating objects: 68713, done.
remote: Counting objects: 100% (229/229), done.
remote: Compressing objects: 100% (131/131), done.
remote: Total 68713 (delta 165), reused 98 (delta 98), pack-reused 68484 (from 4)
Receiving objects: 100% (68713/68713), 203.96 MiB | 13.61 MiB/s, done.
Resolving deltas: 100% (49670/49670), done.
/content/llama.cpp
-- The C compiler identification is GNU 11.4.0
-- The CXX compiler identification is GNU 11.4.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMAKE_BUILD_TYPE=Release
-- Found Git: /usr/bin/git (found version "2.34.1")
-- The ASM compiler identificatio

In [71]:
!./llama.cpp/build/bin/llama-quantize {mtmd} Q4_K_M

main: build = 7150 (3d07caa99)
main: built with cc (Ubuntu 11.4.0-1ubuntu1~22.04.2) 11.4.0 for x86_64-linux-gnu
main: quantizing '/root/.cache/huggingface/hub/models--unsloth--Qwen3-VL-8B-Thinking-GGUF/snapshots/894774171339940a7a7f162a39d8143fb5154cdd/mmproj-BF16.gguf' to '/root/.cache/huggingface/hub/models--unsloth--Qwen3-VL-8B-Thinking-GGUF/snapshots/894774171339940a7a7f162a39d8143fb5154cdd/ggml-model-Q4_K_M.gguf' as Q4_K_M
llama_model_loader: loaded meta data with 31 key-value pairs and 352 tensors from /root/.cache/huggingface/hub/models--unsloth--Qwen3-VL-8B-Thinking-GGUF/snapshots/894774171339940a7a7f162a39d8143fb5154cdd/mmproj-BF16.gguf (version GGUF V3 (latest))
llama_model_loader: Dumping metadata keys/values. Note: KV overrides do not apply in this output.
llama_model_loader: - kv   0:                       general.architecture str              = clip
llama_model_loader: - kv   1:                               general.type str              = mmproj
llama_model_loader: - kv 

In [62]:
model = '/root/.cache/huggingface/hub/models--sentence-transformers--all-MiniLM-L6-v2/snapshots/c9745ed1d9f207416be6d2e6f8de32d1f16199bf/'

In [57]:

# Load model directly
from transformers import AutoTokenizer, AutoModel

tokenizer = AutoTokenizer.from_pretrained("sentence-transformers/all-MiniLM-L6-v2")
model = AutoModel.from_pretrained("sentence-transformers/all-MiniLM-L6-v2")

tokenizer_config.json:   0%|          | 0.00/350 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/612 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

In [52]:
!./llama.cpp/build/bin/llama-cli -t 2 -ngl -1 -c 8192 -sys "You are a Senior AI Engineer!" \
--temp 0.7 --top-p 0.92 --top-k 50 --mirostat 2 --mirostat-lr 2e-4 --batch-size 16 -s 42 -no-cnv \
--log-prefix --color -cmoe -m {model} -p "Generate code snippets to building a model for Agentic RAG system and Vector DB:"

I build: 7150 (3d07caa99) with cc (Ubuntu 11.4.0-1ubuntu1~22.04.2) 11.4.0 for x86_64-linux-gnu
I main: llama backend init
I main: load the model and apply lora adapter, if any
I llama_model_loader: loaded meta data with 35 key-value pairs and 398 tensors from /root/.cache/huggingface/hub/models--mradermacher--Qwen3-4B-Claude-Sonnet-4-Reasoning-Distill-Safetensor-GGUF/snapshots/6862150716160dca2edbd80245e2c5a4ca1a2eac/Qwen3-4B-Claude-Sonnet-4-Reasoning-Distill-Safetensor.Q4_K_M.gguf (version GGUF V3 (latest))
I llama_model_loader: Dumping metadata keys/values. Note: KV overrides do not apply in this output.
I llama_model_loader: - kv   0:                       general.architecture str              = qwen3
I llama_model_loader: - kv   1:                               general.type str              = model
I llama_model_loader: - kv   2:                               general.name str              = Qwen3 4B Claude Sonnet 4 Reasoning Di...
I llama_model_loader: - kv   3:                    

In [49]:
from langchain.embeddings import SentenceTransformerEmbeddings
from langchain.vectorstores import FAISS
from langchain.text_splitter import CharacterTextSplitter
from langchain.docstore.document import Document
import numpy as np

# Example documents
documents = [
    Document(page_content="The quick brown fox jumps over the lazy dog.",
             metadata={"source": "example.txt"}),
    Document(page_content="Vector databases store high-dimensional vectors for efficient similarity search.",
             metadata={"source": "vector_db.txt"}),
    # Add more documents as needed
]

# Split text if needed
text_splitter = CharacterTextSplitter(chunk_size=500, chunk_overlap=50)
docs = text_splitter.split_documents(documents)

# Create embeddings
embedding = SentenceTransformerEmbeddings(model_name=emb)
vectors = embedding.embed_documents([d.page_content for d in docs])

# Create vector store
vector_db = FAISS.from_embeddings(docs, embedding)

ImportError: cannot import name 'SentenceTransformerEmbeddings' from 'langchain.embeddings' (/usr/local/lib/python3.12/dist-packages/langchain/embeddings/__init__.py)

In [44]:
from langchain.agents import AgentExecutor, create_retrieval_agent
from langchain.chains import LLMChain
from langchain.llms import Transformers
from langchain import prompts

# Define prompt template
prompt_template = prompts.load_prompt("retrieval_augment_...")
prompt = prompts.prompt_template(prompt_template)

# Create LLM
llm = Transformers(model_name="gpt-4o-mini")

# Create agent
agent = create_retrieval_agent(llm, vector_db, chain_type="mapreduce")

# Execute agent
response = agent.run("What is the purpose of vector databases?")
print(response)

ImportError: cannot import name 'AgentExecutor' from 'langchain.agents' (/usr/local/lib/python3.12/dist-packages/langchain/agents/__init__.py)

In [40]:
import tensorflow as tf
from keras.models import Sequential
from keras.layers import Dense, Flatten, Conv2D
import numpy as np

# Simple Q-network architecture
class QNetwork(Sequential):
    def __init__(self, state_size, action_size, hidden_units=256, **kwargs):
        super(QNetwork, self).__init__(**kwargs)
        self.state_size = state_size
        self.action_size = action_size

        # First convolutional layer (for image inputs)
        self.add(Conv2D(32, (8,8), strides=(4,4), padding='same', input_shape=state_size))
        # Flatten layer
        self.add(Flatten())
        # Hidden layers
        for _ in range(2):
            self.add(Dense(hidden_units, activation='relu'))
        # Output layer
        self.add(Dense(self.action_size))

    @property
    def action_size(self):
        return self.layers[-1].output_shape[1]

    @property
    def state_size(self):
        return self.layers[0].input_shape[1:]

# Example usage
state_size = (84, 84, 1)  # Standard RGB image for games
action_size = 6
model = QNetwork(state_size, action_size)
model.summary()

IndexError: list index out of range

In [ ]:
from keras.models import Sequential
from keras.layers import Dense, Input
import numpy as np

# Policy network for continuous control
class PolicyNetwork(Sequential):
    def __init__(self, state_size, action_dim, hidden_units=256, **kwargs):
        super(PolicyNetwork, self).__init__(**kwargs)
        self.state_size = state_size
        self.action_dim = action_dim

        # Input layer
        input_layer = Input(shape=state_size)
        # Hidden layers
        x = Dense(hidden_units, activation='relu')(x)
        x = Dense(hidden_units, activation='relu')(x)
        # Output layer - Gaussian distribution parameters
        self.mean = Dense(action_dim)(x)
        self.log_std = Dense(action_dim)(x)

    def call(self, inputs):
        x = inputs
        for layer in self.layers[2:-2]:
            x = layer(x)
        mean = self.mean(x)
        log_std = self.log_std(x)
        # Clip log_std to avoid very small values
        log_std = tf.clip_by_value(log_std, -20, 0)
        # Sample from Gaussian distribution
        sampled_actions = tf.random.normal(tf.shape(mean)) * tf.exp(log_std) + mean
        return sampled_actions, log_std, mean

# Example usage
state_size = (20,)  # Example state space
action_dim = 1
policy_model = PolicyNetwork(state_size, action_dim)
policy_model.summary()

In [48]:
emb = '/root/.cache/huggingface/hub/models--nomic-ai--nomic-embed-text-v2-moe-GGUF/snapshots/ffbcf4c99e5d617dda10ec8c0e9f75754b0cbb80/nomic-embed-text-v2-moe.f32.gguf'

In [34]:
model = '/root/.cache/huggingface/hub/models--mradermacher--Qwen3-4B-Claude-Sonnet-4-Reasoning-Distill-Safetensor-GGUF/snapshots/6862150716160dca2edbd80245e2c5a4ca1a2eac/Qwen3-4B-Claude-Sonnet-4-Reasoning-Distill-Safetensor.Q4_K_M.gguf'

In [70]:
mtmd = '/root/.cache/huggingface/hub/models--unsloth--Qwen3-VL-8B-Thinking-GGUF/snapshots/894774171339940a7a7f162a39d8143fb5154cdd/mmproj-BF16.gguf'

In [74]:

# !pip install llama-cpp-python

from llama_cpp import Llama

llm = Llama.from_pretrained(
    repo_id="unsloth/DeepSeek-R1-Distill-Qwen-1.5B-GGUF",
    filename="DeepSeek-R1-Distill-Qwen-1.5B-Q8_0.gguf",
)

./DeepSeek-R1-Distill-Qwen-1.5B-Q8_0.ggu(…):   0%|          | 0.00/1.89G [00:00<?, ?B/s]

llama_model_loader: loaded meta data with 34 key-value pairs and 339 tensors from /root/.cache/huggingface/hub/models--unsloth--DeepSeek-R1-Distill-Qwen-1.5B-GGUF/snapshots/3cb4d15544a2a5e07439592b9a0965b6445fbd34/./DeepSeek-R1-Distill-Qwen-1.5B-Q8_0.gguf (version GGUF V3 (latest))
llama_model_loader: Dumping metadata keys/values. Note: KV overrides do not apply in this output.
llama_model_loader: - kv   0:                       general.architecture str              = qwen2
llama_model_loader: - kv   1:                               general.type str              = model
llama_model_loader: - kv   2:                               general.name str              = Deepseek-R1-Distill-Qwen-1.5B
llama_model_loader: - kv   3:                           general.basename str              = Deepseek-R1-Distill-Qwen-1.5B
llama_model_loader: - kv   4:                       general.quantized_by str              = Unsloth
llama_model_loader: - kv   5:                         general.size_label str   

In [75]:
llm.create_chat_completion(
  messages = [
    {
      "role": "system",
      "content": "You are a Senior AI Engineer!"
    },
    {
      "role": "user",
      "content": "Create an Python snippets to building a NLP reasoning model?"
    },
  ],
  max_tokens=1024,
  temperature=0.7,
  top_p=0.92,
  top_k=50
)

llama_perf_context_print:        load time =    1512.27 ms
llama_perf_context_print: prompt eval time =    1510.01 ms /    23 tokens (   65.65 ms per token,    15.23 tokens per second)
llama_perf_context_print:        eval time =  109944.72 ms /   488 runs   (  225.30 ms per token,     4.44 tokens per second)
llama_perf_context_print:       total time =  112678.67 ms /   511 tokens
llama_perf_context_print:    graphs reused =        472


{'id': 'chatcmpl-ecd1f92c-d97c-403c-9c14-55035d8beaac',
 'object': 'chat.completion',
 'created': 1764034227,
 'model': '/root/.cache/huggingface/hub/models--unsloth--DeepSeek-R1-Distill-Qwen-1.5B-GGUF/snapshots/3cb4d15544a2a5e07439592b9a0965b6445fbd34/./DeepSeek-R1-Distill-Qwen-1.5B-Q8_0.gguf',
 'choices': [{'index': 0,
   'message': {'role': 'assistant',
    'content': "Alright, the user is asking me to create a Python snippet for building a NLP reasoning model. They mentioned it's for building a reasoning model, so I need to figure out what exactly they're looking for.\n\nFirst, I should consider what a basic NLP model looks like. A common starting point is a sequence-to-sequence model, which can be implemented using PyTorch. The user might be aiming to build something simple, so I should keep it straightforward.\n\nI'll start by outlining the basic components: input, hidden state, and output. For input, a sentence in token form is typical. The hidden state will process this input, 

In [99]:

# Load model directly
from transformers import AutoTokenizer, AutoModelForCausalLM

tokenizer = AutoTokenizer.from_pretrained("mistralai/Mistral-7B-Instruct-v0.2")
model = AutoModelForCausalLM.from_pretrained("mistralai/Mistral-7B-Instruct-v0.2")
messages = [
    {"role": "user", "content": "Who are you?"},
]
inputs = tokenizer.apply_chat_template(
    messages,
    add_generation_prompt=True,
    tokenize=True,
    return_dict=True,
    return_tensors="pt",
).to(model.device)

outputs = model.generate(**inputs, max_new_tokens=4096)
print(tokenizer.decode(outputs[0][inputs["input_ids"].shape[-1]:]))

tokenizer_config.json: 0.00B [00:00, ?B/s]

tokenizer.model:   0%|          | 0.00/493k [00:00<?, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/414 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/596 [00:00<?, ?B/s]

model.safetensors.index.json: 0.00B [00:00, ?B/s]

Fetching 3 files:   0%|          | 0/3 [00:00<?, ?it/s]

model-00003-of-00003.safetensors:   0%|          | 0.00/4.54G [00:00<?, ?B/s]

model-00002-of-00003.safetensors:   0%|          | 0.00/5.00G [00:00<?, ?B/s]

model-00001-of-00003.safetensors:   0%|          | 0.00/4.94G [00:00<?, ?B/s]

KeyboardInterrupt: 

In [89]:
!./llama.cpp/build/bin/llama-cli -s 42 --batch-size 32 \
-t 2 -ngl -1 -c 8192 --temp 0.3 --top-p 0.92 -no-cnv \
-m /root/.cache/huggingface/hub/models--unsloth--DeepSeek-R1-Distill-Qwen-1.5B-GGUF/snapshots/3cb4d15544a2a5e07439592b9a0965b6445fbd34/DeepSeek-R1-Distill-Qwen-1.5B-BF16.gguf \
-p "Generate code snippets to build an artificial intelligence assistant include everything for fine-tuning."

build: 7150 (3d07caa99) with cc (Ubuntu 11.4.0-1ubuntu1~22.04.2) 11.4.0 for x86_64-linux-gnu
main: llama backend init
main: load the model and apply lora adapter, if any
llama_model_loader: loaded meta data with 34 key-value pairs and 339 tensors from /root/.cache/huggingface/hub/models--unsloth--DeepSeek-R1-Distill-Qwen-1.5B-GGUF/snapshots/3cb4d15544a2a5e07439592b9a0965b6445fbd34/DeepSeek-R1-Distill-Qwen-1.5B-BF16.gguf (version GGUF V3 (latest))
llama_model_loader: Dumping metadata keys/values. Note: KV overrides do not apply in this output.
llama_model_loader: - kv   0:                       general.architecture str              = qwen2
llama_model_loader: - kv   1:                               general.type str              = model
llama_model_loader: - kv   2:                               general.name str              = Deepseek-R1-Distill-Qwen-1.5B
llama_model_loader: - kv   3:                           general.basename str              = Deepseek-R1-Distill-Qwen-1.5B
llama_model

<think>
Okay, I need to generate code snippets to build an Artificial Intelligence assistant. Hmm, where do I start? I remember that an AI system typically needs components like a model, data, and a framework. Let me think about each of these parts.

First, the model. I know that models can be machine learning algorithms. Maybe I should include different types like neural networks or decision trees. Oh, and I should explain how the model is trained, so I'll need some code for that. Maybe something like using TensorFlow or PyTorch.

Next, the data. The assistant needs to process and understand the data it receives. So I'll need some code to read from a file or API, parse the data, and maybe preprocess it, like normalizing or encoding categorical variables. That makes sense.

Then there's the framework or library. I think TensorFlow is a good choice because it's widely used and has a lot of libraries. So I'll include importing TensorFlow and maybe TensorFlow.js if I'm targeting a web-based AI assistant.

Training the model is crucial. I'll need code to load the model, load the training data, compile the model with an optimizer and loss function, then fit it to the data. That should train the AI effectively.

Evaluating the model's performance is important too. I'll have to split the data into training and validation sets, use a loss function to measure how well it's performing, and maybe compute metrics like accuracy or loss to see how well it's working.

Deployment is another step. I should explain how to use the AI, maybe through an API. Also, I'll need code to handle requests and make predictions, so I'll include a simple API structure.

Documentation is necessary for anyone to use the AI. So I'll add a docstring explaining the components and what the AI can do. Also, maybe some error handling and logging to make it robust.

I should also think about the architecture. Should it be a single model or multiple models? For now, I'll focus on a single model to keep it simple.

Putting it all together, I'll outline each component step by step, making sure to explain each part clearly. This way, someone can follow along and understand how to build and use an AI assistant.
</think>

To build an Artificial Intelligence assistant, you'll need to design and implement a system that can process information, learn, and make decisions. Below are some code snippets that demonstrate the key components of an AI system:

### 1. Model Component
```python
import tensorflow as tf

# Load and define the model
model = tf.keras.Sequential([tf.keras.layers.Dense(64, activation='relu'), tf.keras.layers.Dense(1, activation='sigmoid')])
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

# Load training data
import numpy as np
train_data = np.array([[0, 0], [0, 1], [1, 0], [1, 1]], dtype=np.float32)
train_labels = np.array([[0, 0], [0, 1], [1, 0], [1, 1]], dtype=np.float32)

# Train the model
model.fit(train_data, train_labels, epochs=500, verbose=2)
```

### 2. Data Component
```python
# Load and prepare the data
import pandas as pd
data = pd.read_csv('data.csv')

# Preprocess the data
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

X = data.drop('target', axis=1).values
y = data['target'].values
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

# Save the data to a file
import pickle
with open('data_painted.pkl', 'wb') as f:
    pickle.dump((X_train, y_train, X_test, y_test), f)
```

### 3. Framework Component
```python
# Import TensorFlow
import tensorflow as tf

# Load the saved model
model = tf.keras.models.load_model('ai_model.h5')

# Make a prediction
new_data = np.array([[2.5]], dtype=np.float32)
 prediction = model.predict(new_data)
 print("Prediction:", prediction)
```

### 4. Training Component
```python
# Import necessary modules
import numpy as np
import matplotlib.pyplot as plt

# Example of training a neural network
# ... (existing training code)

# Plot the training history
plt.plot(history.history['accuracy'], history.history['val_accuracy'])
plt.plot(history.history['loss'], history.history['val_loss'])
plt.xlabel('Training')
plt.ylabel('Loss/Score')
plt.legend(['Accuracy', 'Validation Accuracy', 'Loss', 'Validation Loss'])
plt.show()
```

### 5. Deployment Component
```python
# Example of creating an API
import requests

# API endpoint
url = 'https://localhost:5000/api/predict'
model = 'ai_model'

# Function to make a prediction
def predict_pain(x):
    response = requests.post(url, json={'input': x})
    response.raise_for_status()
    return response.json()['result']

# Make a prediction
input_data = [100]
output = predict_pain(input_data)
print("Pain level:", output)
```

### 6. Documentation
```python
"""
# AI Assistant Documentation

This AI assistant is a simple implementation of a binary classification model that predicts pain levels based on input data.

## Components

### Model
- **Type**: Binary Classification Model
- **Description**: The AI model uses a feedforward neural network with two layers. It takes an input of 1 feature and outputs a prediction between 0 and 1.

### Training
- **Data**: The model is trained on a dataset of 4 examples, each with 1 feature and a binary outcome.
- **Parameters**: The model has 2 weights (one for each layer) and 3 biases.

### Deployment
- **API**: The AI can be used as an API endpoint at http://localhost:5000.
- **Usage**: The API accepts a single input parameter (a float) and returns a binary prediction.

## Usage
1. **Training**: Run the training code to load the model and data.
2. **Testing**: Test the model by passing sample data.
3. **Deployment**: Deploy the model as an API for real-world predictions.

## Limitations
- The model is limited to one feature.
- The API is a frontend and should be tested with actual data.

"""
```

### 7. Error Handling and Logging
```python
# Example of error handling
try:
    input_data = float(input("Enter pain level: "))
    prediction = model.predict([input_data])
    print(f"Prediction: {prediction}")
except Exception as e:
    print(f"Error: {str(e)}")
```

### 8. Conclusion
These code snippets provide a basic implementation of an AI assistant. Depending on your needs, you can expand the architecture by adding more layers, using different algorithms, or integrating additional components like natural language processing or vision-based AI.

warning: no usable GPU found, --gpu-layers option will be ignored
warning: one possible reason is that llama.cpp was compiled without GPU support
warning: consult docs/build.md for compilation instructions
build: 7150 (3d07caa99) with cc (Ubuntu 11.4.0-1ubuntu1~22.04.2) 11.4.0 for x86_64-linux-gnu
main: llama backend init
main: load the model and apply lora adapter, if any
llama_model_loader: loaded meta data with 34 key-value pairs and 339 tensors from /root/.cache/huggingface/hub/models--unsloth--DeepSeek-R1-Distill-Qwen-1.5B-GGUF/snapshots/3cb4d15544a2a5e07439592b9a0965b6445fbd34/DeepSeek-R1-Distill-Qwen-1.5B-BF16.gguf (version GGUF V3 (latest))
llama_model_loader: Dumping metadata keys/values. Note: KV overrides do not apply in this output.
llama_model_loader: - kv   0:                       general.architecture str              = qwen2
llama_model_loader: - kv   1:                               general.type str              = model
llama_model_loader: - kv   2:                               general.name str              = Deepseek-R1-Distill-Qwen-1.5B
llama_model_loader: - kv   3:                           general.basename str              = Deepseek-R1-Distill-Qwen-1.5B
llama_model_loader: - kv   4:                       general.quantized_by str              = Unsloth
llama_model_loader: - kv   5:                         general.size_label str              = 1.5B
llama_model_loader: - kv   6:                            general.license str              = mit
llama_model_loader: - kv   7:                           general.repo_url str              = https://huggingface.co/unsloth
llama_model_loader: - kv   8:                   general.base_model.count u32              = 1
llama_model_loader: - kv   9:                  general.base_model.0.name str              = DeepSeek R1 Distill Qwen 1.5B
llama_model_loader: - kv  10:          general.base_model.0.organization str              = Deepseek Ai
llama_model_loader: - kv  11:              general.base_model.0.repo_url str              = https://huggingface.co/deepseek-ai/De...
llama_model_loader: - kv  12:                               general.tags arr[str,1]       = ["unsloth"]
llama_model_loader: - kv  13:                          qwen2.block_count u32              = 28
llama_model_loader: - kv  14:                       qwen2.context_length u32              = 131072
llama_model_loader: - kv  15:                     qwen2.embedding_length u32              = 1536
llama_model_loader: - kv  16:                  qwen2.feed_forward_length u32              = 8960
llama_model_loader: - kv  17:                 qwen2.attention.head_count u32              = 12
llama_model_loader: - kv  18:              qwen2.attention.head_count_kv u32              = 2
llama_model_loader: - kv  19:                       qwen2.rope.freq_base f32              = 10000.000000
llama_model_loader: - kv  20:     qwen2.attention.layer_norm_rms_epsilon f32              = 0.000001
llama_model_loader: - kv  21:                          general.file_type u32              = 32
llama_model_loader: - kv  22:                       tokenizer.ggml.model str              = gpt2
llama_model_loader: - kv  23:                         tokenizer.ggml.pre str              = deepseek-r1-qwen
llama_model_loader: - kv  24:                      tokenizer.ggml.tokens arr[str,151936]  = ["!", "\"", "#", "$", "%", "&", "'", ...
llama_model_loader: - kv  25:                  tokenizer.ggml.token_type arr[i32,151936]  = [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...
llama_model_loader: - kv  26:                      tokenizer.ggml.merges arr[str,151387]  = ["Ġ Ġ", "ĠĠ ĠĠ", "i n", "Ġ t",...
llama_model_loader: - kv  27:                tokenizer.ggml.bos_token_id u32              = 151646
llama_model_loader: - kv  28:                tokenizer.ggml.eos_token_id u32              = 151643
llama_model_loader: - kv  29:            tokenizer.ggml.padding_token_id u32              = 151643
llama_model_loader: - kv  30:               tokenizer.ggml.add_bos_token bool             = true
llama_model_loader: - kv  31:               tokenizer.ggml.add_eos_token bool             = false
llama_model_loader: - kv  32:                    tokenizer.chat_template str              = {% if not add_generation_prompt is de...
llama_model_loader: - kv  33:               general.quantization_version u32              = 2
llama_model_loader: - type  f32:  141 tensors
llama_model_loader: - type bf16:  198 tensors
print_info: file format = GGUF V3 (latest)
print_info: file type   = BF16
print_info: file size   = 3.31 GiB (16.00 BPW)
load: special_eos_id is not in special_eog_ids - the tokenizer config may be incorrect
load: printing all EOG tokens:
load:   - 151643 ('<｜end▁of▁sentence｜>')
load:   - 151662 ('<|fim_pad|>')
load:   - 151663 ('<|repo_name|>')
load:   - 151664 ('<|file_sep|>')
load: special tokens cache size = 22
load: token to piece cache size = 0.9310 MB
print_info: arch             = qwen2
print_info: vocab_only       = 0
print_info: n_ctx_train      = 131072
print_info: n_embd           = 1536
print_info: n_embd_inp       = 1536
print_info: n_layer          = 28
print_info: n_head           = 12
print_info: n_head_kv        = 2
print_info: n_rot            = 128
print_info: n_swa            = 0
print_info: is_swa_any       = 0
print_info: n_embd_head_k    = 128
print_info: n_embd_head_v    = 128
print_info: n_gqa            = 6
print_info: n_embd_k_gqa     = 256
print_info: n_embd_v_gqa     = 256
print_info: f_norm_eps       = 0.0e+00
print_info: f_norm_rms_eps   = 1.0e-06
print_info: f_clamp_kqv      = 0.0e+00
print_info: f_max_alibi_bias = 0.0e+00
print_info: f_logit_scale    = 0.0e+00
print_info: f_attn_scale     = 0.0e+00
print_info: n_ff             = 8960
print_info: n_expert         = 0
print_info: n_expert_used    = 0
print_info: n_expert_groups  = 0
print_info: n_group_used     = 0
print_info: causal attn      = 1
print_info: pooling type     = -1
print_info: rope type        = 2
print_info: rope scaling     = linear
print_info: freq_base_train  = 10000.0
print_info: freq_scale_train = 1
print_info: n_ctx_orig_yarn  = 131072
print_info: rope_finetuned   = unknown
print_info: model type       = 1.5B
print_info: model params     = 1.78 B
print_info: general.name     = Deepseek-R1-Distill-Qwen-1.5B
print_info: vocab type       = BPE
print_info: n_vocab          = 151936
print_info: n_merges         = 151387
print_info: BOS token        = 151646 '<｜begin▁of▁sentence｜>'
print_info: EOS token        = 151643 '<｜end▁of▁sentence｜>'
print_info: EOT token        = 151643 '<｜end▁of▁sentence｜>'
print_info: PAD token        = 151643 '<｜end▁of▁sentence｜>'
print_info: LF token         = 198 'Ċ'
print_info: FIM PRE token    = 151659 '<|fim_prefix|>'
print_info: FIM SUF token    = 151661 '<|fim_suffix|>'
print_info: FIM MID token    = 151660 '<|fim_middle|>'
print_info: FIM PAD token    = 151662 '<|fim_pad|>'
print_info: FIM REP token    = 151663 '<|repo_name|>'
print_info: FIM SEP token    = 151664 '<|file_sep|>'
print_info: EOG token        = 151643 '<｜end▁of▁sentence｜>'
print_info: EOG token        = 151662 '<|fim_pad|>'
print_info: EOG token        = 151663 '<|repo_name|>'
print_info: EOG token        = 151664 '<|file_sep|>'
print_info: max token length = 256
load_tensors: loading model tensors, this can take a while... (mmap = true)
load_tensors:   CPU_Mapped model buffer size =  3389.80 MiB
............................................................................
llama_context: constructing llama_context
llama_context: n_seq_max     = 1
llama_context: n_ctx         = 8192
llama_context: n_ctx_seq     = 8192
llama_context: n_batch       = 2048
llama_context: n_ubatch      = 512
llama_context: causal_attn   = 1
llama_context: flash_attn    = auto
llama_context: kv_unified    = false
llama_context: freq_base     = 10000.0
llama_context: freq_scale    = 1
llama_context: n_ctx_seq (8192) < n_ctx_train (131072) -- the full capacity of the model will not be utilized
llama_context:        CPU  output buffer size =     0.58 MiB
llama_kv_cache:        CPU KV buffer size =   224.00 MiB
llama_kv_cache: size =  224.00 MiB (  8192 cells,  28 layers,  1/1 seqs), K (f16):  112.00 MiB, V (f16):  112.00 MiB
llama_context: Flash Attention was auto, set to enabled
llama_context:        CPU compute buffer size =   324.76 MiB
llama_context: graph nodes  = 959
llama_context: graph splits = 1
common_init_from_params: added <｜end▁of▁sentence｜> logit bias = -inf
common_init_from_params: added <|fim_pad|> logit bias = -inf
common_init_from_params: added <|repo_name|> logit bias = -inf
common_init_from_params: added <|file_sep|> logit bias = -inf
common_init_from_params: setting dry_penalty_last_n to ctx_size = 8192
common_init_from_params: warming up the model with an empty run - please wait ... (--no-warmup to disable)
main: llama threadpool init, n_threads = 2
main: chat template is available, enabling conversation mode (disable it with -no-cnv)
*** User-specified prompt will pre-start conversation, did you mean to set --system-prompt (-sys) instead?
main: chat template example:
You are a helpful assistant

<｜User｜>Hello<｜Assistant｜>Hi there<｜end▁of▁sentence｜><｜User｜>How are you?<｜Assistant｜>

system_info: n_threads = 2 (n_threads_batch = 2) / 2 | CPU : SSE3 = 1 | SSSE3 = 1 | AVX = 1 | AVX2 = 1 | F16C = 1 | FMA = 1 | BMI2 = 1 | LLAMAFILE = 1 | OPENMP = 1 | REPACK = 1 |

main: interactive mode on.
sampler seed: 2219118661
sampler params:
    repeat_last_n = 64, repeat_penalty = 1.000, frequency_penalty = 0.000, presence_penalty = 0.000
    dry_multiplier = 0.000, dry_base = 1.750, dry_allowed_length = 2, dry_penalty_last_n = 8192
    top_k = 40, top_p = 0.920, min_p = 0.050, xtc_probability = 0.000, xtc_threshold = 0.100, typical_p = 1.000, top_n_sigma = -1.000, temp = 0.700
    mirostat = 0, mirostat_lr = 0.100, mirostat_ent = 5.000
sampler chain: logits -> logit-bias -> penalties -> dry -> top-n-sigma -> top-k -> typical -> top-p -> min-p -> xtc -> temp-ext -> dist
generate: n_ctx = 8192, n_batch = 2048, n_predict = -1, n_keep = 1

== Running in interactive mode. ==
 - Press Ctrl+C to interject at any time.
 - Press Return to return control to the AI.
 - To return control without starting a new line, end your input with '/'.
 - If you want to submit another line, end your input with '\'.
 - Not using system message. To change it, set a different value via -sys PROMPT

Generate code snippets to build an artificial intelligence assistant include everything for fine-tuning.<think>
Okay, so the user asked for code snippets to build an AI assistant that includes everything for fine-tuning. I need to break this down. First, I should think about the main areas involved in AI fine-tuning. That would include language models, training data handling, optimization techniques, and deployment.

Starting with language models, the main ones are GPT, BERT, and others. GPT-3 is widely used, so maybe a snippet for training a GPT model. I remember there's an Adam optimizer, so I'll include that. Also, loss functions like BCEWithLogitsLoss are essential for binary classification tasks.

Next, fine-tuning involves retraining with a specific task. Maybe a snippet that shows how to load a pre-trained model, set the task, and then fine-tune it. Oh, and using a callback to handle early stopping could be useful to prevent overfitting.

Preprocessing is another big part. Text cleaning, tokenization, and label encoding are crucial. For example, removing punctuation and keeping only alphanumeric, then tokenizing with BPE or something like that. Label encoding with one-hot or embeddings can help with model training.

Optimization is next. Adam is a common optimizer, so including that with learning rate scheduling would help. Maybe a learning rate warmup to start with small updates and then decrease.

Evaluation is important too. Metrics like accuracy and loss can help monitor training. Also, saving checkpoints every few epochs would allow for model checkpointing during training.

Deployment is a big part too. Using FastText for text classification, which is efficient. Then, setting up a FastAPI application to handle endpoints and API requests. Also, including Docker for containerization to deploy on multiple platforms.

Now, putting all these together into code snippets. Each snippet should be clear and concise. I'll start with the GPT model training, then move on to fine-tuning setup, preprocessing steps, optimization, evaluation, deployment, and Docker setup.

I should make sure each snippet is self-contained but covers the necessary parts. Including imports and placeholders for variables like the target task and data path will help the user adapt the code to their specific project.

I also need to mention best practices like using proper data augmentation, handling class imbalance, and considering model capacity. That way, the user knows to preprocess their data correctly and adjust their model accordingly.

Finally, I'll wrap it up by suggesting how to integrate all these components into a complete AI assistant. Maybe adding a pipeline that combines data loading, preprocessing, model training, and deployment.
</think>

To build an AI assistant that includes fine-tuning, you'll need to implement several components and code snippets. Below are some examples of code snippets that can help you build an AI assistant with fine-tuning capabilities:

### 1. Language Model Training (GPT Example)

**Code Snippet:**

```python
import torch
import torch.nn as nn
import torch.optim as optim

# Initialize GPT model
model = GPT(n_tokens, n_hiddens, n_layers, n_heads, n Embeddings)
model.load_state_dict(pretrained_model_state_dict)
model.to(device)

# Define optimizer
optimizer = optim.Adam(model.parameters(), lr=learning_rate)

# Define loss function
criterion = nn.CrossEntropyLoss()

# Define training loop
for epoch in range(num_epochs):
    model.train()
    for batch in training_loader:
        optimizer.zero_grad()
        outputs = model(batch)
        loss = criterion(outputs, targets)
        loss.backward()
        optimizer.step()
    
    # Early stopping
    if epoch - best_loss == num_epochs - 1:
        torch.save(model.state_dict(), 'best_model.pth')
        best_loss = loss
        break
```

### 2. Fine-tuning Example

**Code Snippet:**

```python
# Load pre-trained model and task
model = FineTuner(model_name, task_name).to(device)

# Fine-tune the model
model = FineTuner(model_name, task_name).to(device)
criterion = nn.CrossEntropyLoss()
optimizer = optim.Adam(model.parameters(), lr=learning_rate)

# Fine-tune loop
for epoch in range(num_epochs):
    model.train()
    for batch in training_loader:
        # Convert labels to float
        y = torch.tensor(batch['y'], dtype=torch.float)
        # Forward pass
        outputs = model(batch)
        loss = criterion(outputs, y)
        # Backward pass
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()
    
    # Early stopping
    if epoch - best_loss == num_epochs - 1:
        torch.save(model.state_dict(), 'best_model.pth')
        best_loss = loss
        break
```

### 3. Preprocessing Example

**Code Snippet:**

```python
# Preprocess text
text = "Hello, how are you doing?"
tokens = text.lower().split()
# Filter out punctuation and keep only alphanumeric
tokens = [t for t in tokens if t.isalnum()]
# Tokenize using BPE
tokenizer = AutoTokenizer.from_pretrained('bert-base-uncased')
model = tokenizer.encode(' '.join(tokens), add_special_tokens=True)

# Label encoding
labels = ['positive', 'neutral', 'negative']
label_map = {'positive': 0, 'neutral': 1, 'negative': 2}
y = torch.tensor(label_map[t] for t in model)
```

### 4. Optimizer and Learning Rate Scheduling

**Code Snippet:**

```python
# Define optimizer
optimizer = optim.Adam(model.parameters())

# Define learning rate scheduler
scheduler = LRScheduler(optimizer, num_epochs=num_epochs, last_epoch=-1)

# Update learning rate
last_epoch = scheduler.last_epoch
if best_loss == 0:
    scheduler.step(0)
```

### 5. Evaluation

**Code Snippet:**

```python
# Define evaluation metrics
def evaluate(model, val_loader):
    model.eval()
    total = 0
    correct = 0
    with torch.no_grad():
        for batch in val_loader:
            outputs = model(batch)
            _, predicted = torch.max(outputs.data, 1)
            total += len(outputs)
            correct += (predicted == val_labels).sum()
    accuracy = correct / total
    return accuracy

accuracy = evaluate(model, val_loader)
print(f"Accuracy: {accuracy:.4f}")
```

### 6. Deployment Example

**Code Snippet:**

```python
# Set up FastAPI
app = FastAPI()
app.get("/api/predict", response_model=Response)
app.post("/api/predict", response_model=Response)

# Create FastAPI server
@app.get("/api/predict")
def predict_text(text):
    model = load_model('path/to/your/model.pth')
    # Preprocess text
    text = text.strip().lower()
    tokens = text.split()
    if len(tokens) == 0:
        return "0"
    # Encode tokens
    encoded_input = tokenizer.encode(" ".join(tokens), add_special_tokens=True)
    # Make prediction
    with torch.no_grad():
        outputs = model(encoded_input)
    # Get probability of last token
    prob = outputs[-1].max(1)[0].item()
    return prob
```

### 7. Docker Deployment

**Code Snippet:**

```docker
FROM python:3.9-slim
WORKDIR /app
COPY . * .

RUN apt-get update && \
    apt-get install -y \
    build-essential \
    libgl100-dev \
    libgl110-dev \
    libgl120-dev \
    libgl130-dev \
    libgl140-dev \
    libgl150-dev \
    build-essential && \
    docker-compose down

COPY fine-tuner.py . .

RUN apt-get update && \
    apt-get install -y \
    build-essential \
    build-tensorflow . \
    build-tensorflow --no-cache && \
    docker-compose down

COPY data --force --from github.com/kuangliu/transformers \
    --from https://github.com/kuangliu/transformers \
    --to /app/data && \
    docker-compose down && \
    docker-compose up --build
```

### Summary

To build an AI assistant with fine-tuning, you'll need to:

1. Choose a language model architecture (e.g., GPT, BERT).
2. Fine-tune the model on your specific task (e.g., text classification,问答).
3. Implement preprocessing steps (text cleaning, tokenization, label encoding).
4. Use an appropriate loss function and optimizer.
5. Implement early stopping and evaluation metrics.
6. Deploy the model in a production environment (e.g., FastAPI, Docker).
7. Handle deployment and scaling as needed.

You can adapt these code snippets to fit your specific requirements and data.

>

<think>
Okay, so I need to figure out how to generate code snippets that can build an AI assistant. I'm not super experienced with AI, but I know it's related to machine learning and programming. Let me try to break this down step by step.

First, I think I need to understand what an AI assistant does. From what I know, it's supposed to perform tasks that typically require human intelligence, like understanding language, making decisions, or helping with tasks like answering questions. So, building an AI would involve creating systems that can perform these tasks more efficiently.

I remember hearing about machine learning algorithms, so maybe that's a big part of this. It involves training models on data so the AI can learn patterns and make predictions. So, I think I'll need to implement some machine learning concepts in the code.

I also think about the structure of the AI. It should probably have a core system that handles the main tasks and some supporting components like databases or networking. Maybe I'll need to use Python because it's widely used in AI and has libraries like TensorFlow and PyTorch that could help with building models.

Let me outline what I might need to include:

1. **Core System**: This handles the main functionality of the AI assistant. It might include an interface for interacting with the user and other components.

2. **Machine Learning Models**: Implementing algorithms like linear regression, decision trees, or neural networks. Maybe a simple model like a linear regression to predict values.

3. **Data Handling**: Organizing data into training, validation, and testing sets. I think in Python, I can use scikit-learn for splitting datasets and evaluating models.

4. **Network Architecture**: Designing the structure of the neural network. I've heard about Convolutional Neural Networks (CNNs) for image tasks and Recurrent Neural Networks (RNNs) for sequential data. Maybe start with a basic CNN.

5. **Training and Evaluation**: Using libraries to train the model, evaluate its performance, and tweak parameters as needed.

6. **User Interface**: If I'm building a web-based AI assistant, a dashboard or chat interface might be necessary. This would involve using frameworks like Flask or Django.

7. **Integration**: Connecting the AI components together. Maybe using APIs to interact with external systems or services.

Now, let's think about each section in more detail.

**Core System**
- This would be the main part of the AI. It might have an API for making requests and receiving responses. It could handle the user interface, perhaps through web pages or a mobile app.

**Machine Learning Models**
- I'll need to import libraries like scikit-learn. I can create a simple model, like a linear regression that takes user input and predicts a value. For example, predicting house prices based on square footage.

**Data Handling**
- Use scikit-learn to split data into training, validation, and test sets. Maybe use train_test_split for that. I'll need to organize my data into a pandas DataFrame and split it accordingly.

**Network Architecture**
- I'll design a simple CNN using Keras or TensorFlow. Maybe one layer with CNN, another with a fully connected layer, and a final layer for output.

**Training and Evaluation**
- Use fit and evaluate methods in Keras or TensorFlow to train the model. Calculate metrics like accuracy or loss to assess performance and adjust hyperparameters if needed.

**User Interface**
- For a web-based assistant, Flask could be used to serve static files and handle HTTP requests. I can create a Flask app that serves a welcome page and an API endpoint.

**Integration**
- Use requests library to send HTTP requests. Maybe connect the AI model to an external API that provides data or instructions.

Now, putting this all together, I'll need to outline the code snippets. Each section will be a separate code block within a larger script.

Wait, but the user asked for code snippets, so each part should be a separate section. I'll need to create functions or classes for each component. For example, a Model class for the machine learning model, a Network class for the neural network, and a Core class for the AI core system.

I should also consider using best practices in code organization, like creating modules for each component and using proper imports. Maybe include comments to explain what each part does.

I'm a bit unsure about the exact steps for implementing a simple model. For linear regression, I'll need to set up the model, fit it to the training data, and then test it on the test set. Also, handling edge cases, like when the model doesn't perform well, and how to improve it.

For the network architecture, I'll need to define layers with appropriate inputs and outputs. Maybe a simple CNN that takes images and outputs predictions.

In the user interface, using Flask's setup functions like app.run() and get started() will make it easy to start the server and access the API.

I think I've covered the main areas. Now, I'll structure the code snippets accordingly, making sure each part is clear and self-contained within the larger script.
</think>

To create an AI assistant, we'll structure the code into key components, each handling a specific part of the AI system. Below is a detailed breakdown of how each component can be implemented in Python using standard libraries.

### 1. Core System

**API Implementation**
```python
from flask import Flask, render_template_string
from flask import request, jsonify, session
from flask import jsonify

app = Flask(__name__)
app.config['SESSION_COOKIE_SECURE'] = True

@app.route('/')
def home():
    return render_template('index.html')

@app.route('/start')
def start():
    return jsonify({'status': 'success'})
```

**User Interface**
```python
from flask import sendFile
import os
import time

@app.route('/start', methods=['POST'])
def serve_file():
    file = request.files.get('file')
    if file:
        filename = 'assistant.html'
        file.save(os.path.join(__file__, filename))
        sendFile(filename)
    return jsonify({'status': 'success'})
```

### 2. Machine Learning Models

**Linear Regression**
```python
class LinearRegression:
    def __init__(self):
        self.model = None
    
    def fit(self, X, y):
        from sklearn.linear_model import LinearRegression
        self.model = LinearRegression()
        self.model.fit(X, y)
    
    def predict(self, X):
        return self.model.predict(X)
```

### 3. Data Handling

**Data Splitting**
```python
from sklearn.model_selection import train_test_split
import pandas as pd

# Assuming 'data' is a DataFrame
X = data.drop('target', axis=1)
y = data['target']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)
```

### 4. Network Architecture

```python
import tensorflow as tf

class SimpleCNN(tf.keras.Model):
    def __init__(self, input_shape):
        super(SimpleCNN, self).__init__()
        self.model = tf.keras.Sequential([
            tf.keras.layers.Conv2D(32, (3, 3), activation='relu', input_shape=input_shape),
            tf.keras.layers.MaxPooling2D(),
            tf.keras.layers.Flatten(),
            tf.keras.layers.Dense(10, activation='softmax')
        ])
```

### 5. Training and Evaluation

```python
import numpy as np
from sklearn import metrics

# Example training loop
model = SimpleCNN(input_shape=[64, 64, 3])
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model.fit(X_train, y_train, epochs=10, validation_data=(X_test, y_test))
# Evaluate and improve
print("Final Model Metrics:", metrics.classification_report(y_test, model.predict(X_test)))
```

### 6. User Interface Integration

```python
from flask import setup, run, shutdown, request, abort, jsonify
from flask import redirect, url_for
from flask_app import app

app = setup()

@app.route('/api', methods=['GET', 'POST'])
def api():
    if request.method == 'POST':
        # Process request
        result = model.predict([image])
        return jsonify(result)
    return jsonify({'status': 'success'})

if __name__ == '__main__':
    run()
    app.run()
    setup()
```

### Conclusion

This code provides a basic structure for an AI assistant using Flask as the frontend and TensorFlow for neural networks. Each component is designed to be modular and self-contained, allowing for easier integration and extension.

# Task
Prepare a dummy dataset for text classification, use the `llm` object to generate predictions for the texts in the dataset, implement evaluation metrics (accuracy, precision, recall, F1-score) by comparing the model's predictions against the true labels, and display the evaluation results.

## Prepare a Sample Dataset for Evaluation

### Subtask:
Create a small, dummy dataset with input texts and corresponding labels to serve as evaluation data. This will allow us to simulate a specific task, such as simple text classification.


**Reasoning**:
To prepare a sample dataset for evaluation, I will create a Python list of dictionaries named `evaluation_data` with 'text' and 'label' keys, as specified in the instructions.



In [100]:
evaluation_data = [
    {'text': 'This is a great movie!', 'label': 'positive'},
    {'text': 'I hated the food.', 'label': 'negative'},
    {'text': 'The software update improved performance.', 'label': 'IT'},
    {'text': 'The new policy passed in parliament.', 'label': 'Politics'},
    {'text': 'Excellent customer service!', 'label': 'positive'},
    {'text': 'The system crashed again.', 'label': 'IT'},
    {'text': 'Political debate heated up.', 'label': 'Politics'},
    {'text': 'Worst experience ever.', 'label': 'negative'}
]

print(f"Created evaluation_data with {len(evaluation_data)} samples.")

Created evaluation_data with 8 samples.


**Reasoning**:
I need to use the `llm` object, which is an instance of `llama_cpp.Llama`, to generate predictions for each text in the `evaluation_data`. I will iterate through the dataset, craft a suitable prompt for classification, call the `llm.create_chat_completion` method, and store the model's predicted labels. I will also print the text, true label and predicted label.



In [101]:
model_predictions = []

for item in evaluation_data:
    user_text = item['text']
    true_label = item['label']

    # Craft a prompt for text classification
    messages = [
        {"role": "system", "content": "You are a text classification assistant. Classify the following text into one of these categories: 'positive', 'negative', 'IT', 'Politics'. Respond only with the category name."},
        {"role": "user", "content": f"Classify this text: '{user_text}'"},
    ]

    try:
        response = llm.create_chat_completion(
            messages=messages,
            max_tokens=10, # Keep max_tokens low as we only expect a single word label
            temperature=0.0, # Use low temperature for deterministic output
            top_p=1.0,
            top_k=1,
            stop=['.', '\n'] # Stop generation after the first word or newline
        )
        predicted_label = response['choices'][0]['message']['content'].strip()
        model_predictions.append(predicted_label)
        print(f"Text: '{user_text}' | True Label: '{true_label}' | Predicted Label: '{predicted_label}'")
    except Exception as e:
        print(f"Error processing text '{user_text}': {e}")
        model_predictions.append("ERROR") # Append a placeholder for errors

print("\nGenerated predictions for all samples.")

Llama.generate: 4 prefix-match hit, remaining 50 prompt tokens to eval
llama_perf_context_print:        load time =    1512.27 ms
llama_perf_context_print: prompt eval time =   12435.81 ms /    50 tokens (  248.72 ms per token,     4.02 tokens per second)
llama_perf_context_print:        eval time =    2139.75 ms /     9 runs   (  237.75 ms per token,     4.21 tokens per second)
llama_perf_context_print:       total time =   14613.03 ms /    59 tokens
llama_perf_context_print:    graphs reused =          8
Llama.generate: 45 prefix-match hit, remaining 8 prompt tokens to eval


Text: 'This is a great movie!' | True Label: 'positive' | Predicted Label: 'Okay, so I need to classify the text "'


llama_perf_context_print:        load time =    1512.27 ms
llama_perf_context_print: prompt eval time =     517.24 ms /     8 tokens (   64.66 ms per token,    15.47 tokens per second)
llama_perf_context_print:        eval time =    1864.60 ms /     9 runs   (  207.18 ms per token,     4.83 tokens per second)
llama_perf_context_print:       total time =    2399.35 ms /    17 tokens
llama_perf_context_print:    graphs reused =          8
Llama.generate: 45 prefix-match hit, remaining 9 prompt tokens to eval


Text: 'I hated the food.' | True Label: 'negative' | Predicted Label: 'Okay, so I need to classify the text "'


llama_perf_context_print:        load time =    1512.27 ms
llama_perf_context_print: prompt eval time =     673.00 ms /     9 tokens (   74.78 ms per token,    13.37 tokens per second)
llama_perf_context_print:        eval time =    1903.30 ms /     9 runs   (  211.48 ms per token,     4.73 tokens per second)
llama_perf_context_print:       total time =    2594.35 ms /    18 tokens
llama_perf_context_print:    graphs reused =          8
Llama.generate: 46 prefix-match hit, remaining 9 prompt tokens to eval


Text: 'The software update improved performance.' | True Label: 'IT' | Predicted Label: 'Okay, so I need to classify the text ''


llama_perf_context_print:        load time =    1512.27 ms
llama_perf_context_print: prompt eval time =     651.45 ms /     9 tokens (   72.38 ms per token,    13.82 tokens per second)
llama_perf_context_print:        eval time =    1845.00 ms /     9 runs   (  205.00 ms per token,     4.88 tokens per second)
llama_perf_context_print:       total time =    2514.07 ms /    18 tokens
llama_perf_context_print:    graphs reused =          8
Llama.generate: 45 prefix-match hit, remaining 7 prompt tokens to eval


Text: 'The new policy passed in parliament.' | True Label: 'Politics' | Predicted Label: 'Okay, so I need to classify the text "'


llama_perf_context_print:        load time =    1512.27 ms
llama_perf_context_print: prompt eval time =     587.97 ms /     7 tokens (   84.00 ms per token,    11.91 tokens per second)
llama_perf_context_print:        eval time =    2732.91 ms /     9 runs   (  303.66 ms per token,     3.29 tokens per second)
llama_perf_context_print:       total time =    3344.89 ms /    16 tokens
llama_perf_context_print:    graphs reused =          8
Llama.generate: 45 prefix-match hit, remaining 8 prompt tokens to eval


Text: 'Excellent customer service!' | True Label: 'positive' | Predicted Label: 'Okay, so I need to classify the text ''


llama_perf_context_print:        load time =    1512.27 ms
llama_perf_context_print: prompt eval time =     920.37 ms /     8 tokens (  115.05 ms per token,     8.69 tokens per second)
llama_perf_context_print:        eval time =    1870.72 ms /     9 runs   (  207.86 ms per token,     4.81 tokens per second)
llama_perf_context_print:       total time =    2808.47 ms /    17 tokens
llama_perf_context_print:    graphs reused =          8
Llama.generate: 45 prefix-match hit, remaining 8 prompt tokens to eval


Text: 'The system crashed again.' | True Label: 'IT' | Predicted Label: 'Okay, so I need to classify the text ''


llama_perf_context_print:        load time =    1512.27 ms
llama_perf_context_print: prompt eval time =     514.13 ms /     8 tokens (   64.27 ms per token,    15.56 tokens per second)
llama_perf_context_print:        eval time =    1864.07 ms /     9 runs   (  207.12 ms per token,     4.83 tokens per second)
llama_perf_context_print:       total time =    2396.95 ms /    17 tokens
llama_perf_context_print:    graphs reused =          8
Llama.generate: 45 prefix-match hit, remaining 8 prompt tokens to eval


Text: 'Political debate heated up.' | True Label: 'Politics' | Predicted Label: 'Okay, so I need to classify the text ''


llama_perf_context_print:        load time =    1512.27 ms
llama_perf_context_print: prompt eval time =     551.07 ms /     8 tokens (   68.88 ms per token,    14.52 tokens per second)
llama_perf_context_print:        eval time =    1913.66 ms /     9 runs   (  212.63 ms per token,     4.70 tokens per second)
llama_perf_context_print:       total time =    2482.40 ms /    17 tokens
llama_perf_context_print:    graphs reused =          8


Text: 'Worst experience ever.' | True Label: 'negative' | Predicted Label: 'Okay, so I need to classify the text "'

Generated predictions for all samples.


**Reasoning**:
The previous attempt failed because the model was not providing just the category name. I will refine the prompt to be more direct about the expected output format and also implement a more robust extraction of the predicted label from the response content to ensure only the category name is captured. Also, some predicted labels could potentially contain special characters, so I will strip them off. The output is then formatted to lowercase to avoid case-sensitivity issues, so it can later be compared to the true labels.



In [102]:
model_predictions = []
true_labels = []

for item in evaluation_data:
    user_text = item['text']
    true_label = item['label']
    true_labels.append(true_label)

    # Craft a prompt for text classification, emphasizing strict output format
    messages = [
        {"role": "system", "content": "You are a text classification assistant. Classify the following text into one of these categories: 'positive', 'negative', 'IT', 'Politics'. Your response MUST be ONLY the category name, nothing else. For example, if the category is 'positive', respond 'positive'."},
        {"role": "user", "content": f"Classify this text: '{user_text}'"}
    ]

    try:
        response = llm.create_chat_completion(
            messages=messages,
            max_tokens=10, # Keep max_tokens low as we only expect a single word label
            temperature=0.0, # Use low temperature for deterministic output
            top_p=1.0,
            top_k=1,
            stop=['.', '\n', ' ']
        )

        # Extract and clean the predicted label
        predicted_content = response['choices'][0]['message']['content'].strip()
        # Take only the first word and clean any non-alphanumeric characters
        predicted_label = ''.join(filter(str.isalnum, predicted_content.split(' ')[0])).lower()

        model_predictions.append(predicted_label)
        print(f"Text: '{user_text}' | True Label: '{true_label}' | Predicted Label: '{predicted_label}'")
    except Exception as e:
        print(f"Error processing text '{user_text}': {e}")
        model_predictions.append("ERROR") # Append a placeholder for errors

print("\nGenerated predictions for all samples.")

Llama.generate: 31 prefix-match hit, remaining 42 prompt tokens to eval
llama_perf_context_print:        load time =    1512.27 ms
llama_perf_context_print: prompt eval time =    3692.50 ms /    42 tokens (   87.92 ms per token,    11.37 tokens per second)
llama_perf_context_print:        eval time =     496.74 ms /     2 runs   (  248.37 ms per token,     4.03 tokens per second)
llama_perf_context_print:       total time =    4197.53 ms /    44 tokens
llama_perf_context_print:    graphs reused =          1
Llama.generate: 64 prefix-match hit, remaining 8 prompt tokens to eval


Text: 'This is a great movie!' | True Label: 'positive' | Predicted Label: 'okay'


llama_perf_context_print:        load time =    1512.27 ms
llama_perf_context_print: prompt eval time =     535.88 ms /     8 tokens (   66.99 ms per token,    14.93 tokens per second)
llama_perf_context_print:        eval time =     413.70 ms /     2 runs   (  206.85 ms per token,     4.83 tokens per second)
llama_perf_context_print:       total time =     955.31 ms /    10 tokens
llama_perf_context_print:    graphs reused =          1
Llama.generate: 64 prefix-match hit, remaining 9 prompt tokens to eval


Text: 'I hated the food.' | True Label: 'negative' | Predicted Label: 'okay'


llama_perf_context_print:        load time =    1512.27 ms
llama_perf_context_print: prompt eval time =     663.79 ms /     9 tokens (   73.75 ms per token,    13.56 tokens per second)
llama_perf_context_print:        eval time =     406.31 ms /     2 runs   (  203.15 ms per token,     4.92 tokens per second)
llama_perf_context_print:       total time =    1075.70 ms /    11 tokens
llama_perf_context_print:    graphs reused =          1
Llama.generate: 65 prefix-match hit, remaining 9 prompt tokens to eval


Text: 'The software update improved performance.' | True Label: 'IT' | Predicted Label: 'okay'


llama_perf_context_print:        load time =    1512.27 ms
llama_perf_context_print: prompt eval time =     659.67 ms /     9 tokens (   73.30 ms per token,    13.64 tokens per second)
llama_perf_context_print:        eval time =     421.51 ms /     2 runs   (  210.75 ms per token,     4.74 tokens per second)
llama_perf_context_print:       total time =    1087.91 ms /    11 tokens
llama_perf_context_print:    graphs reused =          1
Llama.generate: 64 prefix-match hit, remaining 7 prompt tokens to eval


Text: 'The new policy passed in parliament.' | True Label: 'Politics' | Predicted Label: 'okay'


llama_perf_context_print:        load time =    1512.27 ms
llama_perf_context_print: prompt eval time =     556.92 ms /     7 tokens (   79.56 ms per token,    12.57 tokens per second)
llama_perf_context_print:        eval time =     430.79 ms /     2 runs   (  215.39 ms per token,     4.64 tokens per second)
llama_perf_context_print:       total time =     993.51 ms /     9 tokens
llama_perf_context_print:    graphs reused =          1
Llama.generate: 64 prefix-match hit, remaining 8 prompt tokens to eval


Text: 'Excellent customer service!' | True Label: 'positive' | Predicted Label: 'okay'


llama_perf_context_print:        load time =    1512.27 ms
llama_perf_context_print: prompt eval time =     533.37 ms /     8 tokens (   66.67 ms per token,    15.00 tokens per second)
llama_perf_context_print:        eval time =     404.69 ms /     2 runs   (  202.35 ms per token,     4.94 tokens per second)
llama_perf_context_print:       total time =     943.48 ms /    10 tokens
llama_perf_context_print:    graphs reused =          1
Llama.generate: 64 prefix-match hit, remaining 8 prompt tokens to eval


Text: 'The system crashed again.' | True Label: 'IT' | Predicted Label: 'okay'


llama_perf_context_print:        load time =    1512.27 ms
llama_perf_context_print: prompt eval time =     528.53 ms /     8 tokens (   66.07 ms per token,    15.14 tokens per second)
llama_perf_context_print:        eval time =     406.40 ms /     2 runs   (  203.20 ms per token,     4.92 tokens per second)
llama_perf_context_print:       total time =     940.68 ms /    10 tokens
llama_perf_context_print:    graphs reused =          1
Llama.generate: 64 prefix-match hit, remaining 8 prompt tokens to eval


Text: 'Political debate heated up.' | True Label: 'Politics' | Predicted Label: 'okay'


llama_perf_context_print:        load time =    1512.27 ms
llama_perf_context_print: prompt eval time =     512.91 ms /     8 tokens (   64.11 ms per token,    15.60 tokens per second)
llama_perf_context_print:        eval time =     421.28 ms /     2 runs   (  210.64 ms per token,     4.75 tokens per second)
llama_perf_context_print:       total time =     939.90 ms /    10 tokens
llama_perf_context_print:    graphs reused =          1


Text: 'Worst experience ever.' | True Label: 'negative' | Predicted Label: 'okay'

Generated predictions for all samples.


**Reasoning**:
The model is not adhering to the strict output format and is instead providing conversational responses. I will modify the `llm.create_chat_completion` call to use `response_format` to enforce JSON output for the predicted category, which is a more robust way to ensure structured responses. I will then parse the JSON output to extract the predicted label.



In [103]:
model_predictions = []
true_labels = []

for item in evaluation_data:
    user_text = item['text']
    true_label = item['label']
    true_labels.append(true_label)

    # Craft a prompt for text classification, emphasizing strict output format
    messages = [
        {"role": "system", "content": "You are a text classification assistant. Classify the following text into one of these categories: 'positive', 'negative', 'IT', 'Politics'. Your response MUST be a JSON object with a single key 'category' and its value being the category name. Example: {\"category\": \"positive\"}."},
        {"role": "user", "content": f"Classify this text: '{user_text}'"}
    ]

    try:
        response = llm.create_chat_completion(
            messages=messages,
            max_tokens=20, # Increased max_tokens slightly to accommodate JSON output
            temperature=0.0, # Use low temperature for deterministic output
            top_p=1.0,
            top_k=1,
            response_format={ "type": "json_object" }
        )

        # Extract and clean the predicted label from JSON
        import json
        predicted_json = json.loads(response['choices'][0]['message']['content'])
        predicted_label = predicted_json.get('category', '').lower().strip()

        model_predictions.append(predicted_label)
        print(f"Text: '{user_text}' | True Label: '{true_label}' | Predicted Label: '{predicted_label}'")
    except Exception as e:
        print(f"Error processing text '{user_text}': {e}")
        model_predictions.append("ERROR") # Append a placeholder for errors

print("\nGenerated predictions for all samples.")

Llama.generate: 35 prefix-match hit, remaining 43 prompt tokens to eval
llama_perf_context_print:        load time =    1512.27 ms
llama_perf_context_print: prompt eval time =    3980.32 ms /    43 tokens (   92.57 ms per token,    10.80 tokens per second)
llama_perf_context_print:        eval time =    1884.41 ms /     9 runs   (  209.38 ms per token,     4.78 tokens per second)
llama_perf_context_print:       total time =    6635.02 ms /    52 tokens
llama_perf_context_print:    graphs reused =          8
Llama.generate: 69 prefix-match hit, remaining 8 prompt tokens to eval


Text: 'This is a great movie!' | True Label: 'positive' | Predicted Label: 'positive'


llama_perf_context_print:        load time =    1512.27 ms
llama_perf_context_print: prompt eval time =     510.71 ms /     8 tokens (   63.84 ms per token,    15.66 tokens per second)
llama_perf_context_print:        eval time =    1258.63 ms /     6 runs   (  209.77 ms per token,     4.77 tokens per second)
llama_perf_context_print:       total time =    2250.13 ms /    14 tokens
llama_perf_context_print:    graphs reused =          5
Llama.generate: 69 prefix-match hit, remaining 9 prompt tokens to eval


Text: 'I hated the food.' | True Label: 'negative' | Predicted Label: 'negative'


llama_perf_context_print:        load time =    1512.27 ms
llama_perf_context_print: prompt eval time =     633.00 ms /     9 tokens (   70.33 ms per token,    14.22 tokens per second)
llama_perf_context_print:        eval time =    1831.00 ms /     9 runs   (  203.44 ms per token,     4.92 tokens per second)
llama_perf_context_print:       total time =    3110.06 ms /    18 tokens
llama_perf_context_print:    graphs reused =          8
Llama.generate: 70 prefix-match hit, remaining 9 prompt tokens to eval


Text: 'The software update improved performance.' | True Label: 'IT' | Predicted Label: 'positive'


llama_perf_context_print:        load time =    1512.27 ms
llama_perf_context_print: prompt eval time =     639.87 ms /     9 tokens (   71.10 ms per token,    14.07 tokens per second)
llama_perf_context_print:        eval time =    2273.56 ms /     9 runs   (  252.62 ms per token,     3.96 tokens per second)
llama_perf_context_print:       total time =    3806.61 ms /    18 tokens
llama_perf_context_print:    graphs reused =          8
Llama.generate: 69 prefix-match hit, remaining 7 prompt tokens to eval


Text: 'The new policy passed in parliament.' | True Label: 'Politics' | Predicted Label: 'positive'


llama_perf_context_print:        load time =    1512.27 ms
llama_perf_context_print: prompt eval time =     558.17 ms /     7 tokens (   79.74 ms per token,    12.54 tokens per second)
llama_perf_context_print:        eval time =    1850.56 ms /     9 runs   (  205.62 ms per token,     4.86 tokens per second)
llama_perf_context_print:       total time =    3050.05 ms /    16 tokens
llama_perf_context_print:    graphs reused =          8
Llama.generate: 69 prefix-match hit, remaining 8 prompt tokens to eval


Text: 'Excellent customer service!' | True Label: 'positive' | Predicted Label: 'positive'


llama_perf_context_print:        load time =    1512.27 ms
llama_perf_context_print: prompt eval time =     534.71 ms /     8 tokens (   66.84 ms per token,    14.96 tokens per second)
llama_perf_context_print:        eval time =    1841.75 ms /     9 runs   (  204.64 ms per token,     4.89 tokens per second)
llama_perf_context_print:       total time =    3027.82 ms /    17 tokens
llama_perf_context_print:    graphs reused =          8
Llama.generate: 69 prefix-match hit, remaining 8 prompt tokens to eval


Text: 'The system crashed again.' | True Label: 'IT' | Predicted Label: 'negative'


llama_perf_context_print:        load time =    1512.27 ms
llama_perf_context_print: prompt eval time =     530.12 ms /     8 tokens (   66.27 ms per token,    15.09 tokens per second)
llama_perf_context_print:        eval time =    1866.47 ms /     9 runs   (  207.39 ms per token,     4.82 tokens per second)
llama_perf_context_print:       total time =    3042.59 ms /    17 tokens
llama_perf_context_print:    graphs reused =          8
Llama.generate: 69 prefix-match hit, remaining 8 prompt tokens to eval


Text: 'Political debate heated up.' | True Label: 'Politics' | Predicted Label: 'politics'


llama_perf_context_print:        load time =    1512.27 ms
llama_perf_context_print: prompt eval time =     537.66 ms /     8 tokens (   67.21 ms per token,    14.88 tokens per second)
llama_perf_context_print:        eval time =    2277.71 ms /     9 runs   (  253.08 ms per token,     3.95 tokens per second)
llama_perf_context_print:       total time =    3728.49 ms /    17 tokens
llama_perf_context_print:    graphs reused =          8


Text: 'Worst experience ever.' | True Label: 'negative' | Predicted Label: 'negative'

Generated predictions for all samples.


**Reasoning**:
Now that the model has generated predictions, I will calculate common classification metrics (accuracy, precision, recall, and F1-score) to evaluate its performance against the true labels. I will use `sklearn.metrics.classification_report` for a comprehensive summary.



In [104]:
from sklearn.metrics import classification_report

# Ensure all labels are lowercase for consistent comparison
true_labels_lower = [label.lower() for label in true_labels]
model_predictions_lower = [pred.lower() for pred in model_predictions]

# Filter out any 'error' predictions if they occurred
# We'll create filtered lists to avoid affecting metric calculation
filtered_true_labels = []
filtered_model_predictions = []

for i in range(len(true_labels_lower)):
    if model_predictions_lower[i] != "error":
        filtered_true_labels.append(true_labels_lower[i])
        filtered_model_predictions.append(model_predictions_lower[i])

# Print the classification report
print("\n--- Classification Report ---")
print(classification_report(filtered_true_labels, filtered_model_predictions, zero_division=0))


--- Classification Report ---
              precision    recall  f1-score   support

          it       0.00      0.00      0.00         2
    negative       0.67      1.00      0.80         2
    politics       1.00      0.50      0.67         2
    positive       0.50      1.00      0.67         2

    accuracy                           0.62         8
   macro avg       0.54      0.62      0.53         8
weighted avg       0.54      0.62      0.53         8



## Display Evaluation Results

### Subtask:
Present the calculated evaluation metrics clearly, explaining what each metric indicates about the model's performance on the simulated task.


## Summary:

### Data Analysis Key Findings

*   A dummy dataset comprising 8 samples for text classification was successfully created.
*   Initial attempts to obtain predictions from the LLM resulted in conversational responses rather than structured labels, highlighting the need for explicit output formatting.
*   Enforcing JSON output format for the LLM's predictions via `response_format={"type": "json_object"}` successfully yielded parseable structured responses, extracting category names from the "category" key.
*   The model achieved an overall accuracy of 0.62 on the dummy dataset.
*   Performance varied significantly across categories: 'negative' and 'positive' labels showed better performance, while the 'IT' category had very poor results with a precision, recall, and F1-score of 0.00. The 'Politics' category had a recall of 0.50.

### Insights or Next Steps

*   To improve the model's classification accuracy, especially for underperforming categories like 'IT', refine the prompt to provide clearer instructions, more context, or examples (few-shot learning).
*   Investigate the specific misclassifications for the 'IT' and 'Politics' categories to understand why the LLM struggled with them, potentially by examining the LLM's raw outputs for those samples.


In [106]:
!ls /root/.cache/huggingface/hub/models--unsloth--DeepSeek-R1-Distill-Qwen

models--google--electra-base-discriminator
models--mistralai--Mistral-7B-Instruct-v0.2
models--mradermacher--falcon-7b-instruct-GGUF
models--mradermacher--Qwen3-4B-Claude-Sonnet-4-Reasoning-Distill-Safetensor-GGUF
models--nomic-ai--nomic-embed-text-v2-moe-GGUF
models--sentence-transformers--all-MiniLM-L6-v2
models--unsloth--DeepSeek-R1-Distill-Qwen-1.5B-GGUF
models--unsloth--Qwen3-VL-8B-Thinking-GGUF
